examples/long_running_measurement/measurement.py

```diff
@@ -0,0 +1,83 @@
+"""Perform a long running measurement."""
+import logging
+import pathlib
+import random
+import threading
+import time
+from typing import Generator, List, Tuple
+
+import click
+import grpc
+from _helpers import configure_logging, verbosity_option
+
+import ni_measurementlink_service as nims
+
+service_directory = pathlib.Path(__file__).resolve().parent
+measurement_service = nims.MeasurementService(
+    service_config_path=service_directory / "LongRunningMeasurement.serviceconfig",
+    version="0.5.0.0",
+    ui_file_paths=[service_directory / "LongRunningMeasurement.measui"],
+)
+
+
+Outputs = Tuple[List[int], str, List[int], float]
+
+
+@measurement_service.register_measurement
+@measurement_service.configuration("time_in_seconds", nims.DataType.Int32, 10)
+@measurement_service.output("random_numbers", nims.DataType.Int32Array1D)
+@measurement_service.output("status", nims.DataType.String)
+@measurement_service.output("overlimit", nims.DataType.Int32Array1D)
+@measurement_service.output("elapsed_time", nims.DataType.Double)
+def measure(time_in_seconds: float) -> Generator[Outputs, None, Outputs]:
+    """Perform a long running measurement."""
+    logging.info("Executing measurement: time_in_seconds=%d", time_in_seconds)
+    cancellation_event = threading.Event()
+    measurement_service.context.add_cancel_callback(cancellation_event.set)
+
+    status = ""
+    random_numbers: List[int] = []
+    overlimit: List[int] = []
+    elapsed_time = 0.0
+
+    start_time = time.monotonic()
+    stop_time = start_time + time_in_seconds
+    loop_index = 0
+    while time.monotonic() < stop_time:
+        status = f"Loop index: {loop_index}"
+        loop_index += 1
+
+        elapsed_time = time.monotonic() - start_time
+
+        desired_len = int(elapsed_time * 1000.0 / 10.0)
+        while len(random_numbers) < desired_len:
+            random_number = random.randint(1, 1000)
+            random_numbers.append(random_number)
+            if random_number > 800:
+                overlimit.append(random_number)
+
+        yield (random_numbers, status, overlimit, elapsed_time)
+
+        if cancellation_event.wait(10e-3):
+            logging.info("Canceling measurement")
+            measurement_service.context.abort(
+                grpc.StatusCode.CANCELLED, "Client requested cancellation"
+            )
+
+    logging.info("Completed measurement")
+    status = "Measurement complete."
+    return (random_numbers, status, overlimit, elapsed_time)
```

Comment:
It seems to me like this last set of outputs returned would always be the same as the last yield? That's not necessarily a problem, definitely not a big problem. But I'm curious what the rules are here. Does the measure method have to have a return at the end? Does it have to yield at least once if there isn't a return statement?

Reply:
> It seems to me like this last set of outputs returned would always be the same as the last yield?

The `status` field is different.

> But I'm curious what the rules are here. Does the measure method have to have a return at the end?

No. The `return` is optional. Normally, a generator function in Python only uses `yield`. However, it would be confusing if adding a `yield` to your measurement function caused the `return` to be ignored.

> Does it have to yield at least once if there isn't a return statement?

Yes, it must yield or return a result.

Reply:
I added comments explaining this.